### PR TITLE
Added configDirectory parameter to load yaml config during launch

### PR DESCRIPTION
**Pull request by maheshkelkar**
Added a new parameter, iterating through directories and loading configuration,
configuration validation and sanitization, etc

References #45

### CODE REVIEW

File: deploydb.cucumber.yml [hunk 1]
```diff
@@ -2,6 +2,12 @@ database:
   driverClass: org.h2.Driver
   user: nobody
   url: "jdbc:h2:mem:cucumber"
+
+# DeployDB base configuration directory (cannonical path)
+# YAML configuration files for models - services, environments, pipelines,
+# webhooks and promotions - should be stored in respective sub-directories
+#configDirectory: config
+
```
Comment (GiriDandu) [inline deploydb.cucumber.yml]: Why do you have this commented out. 
You want to remove it from this pull request?

File: src/main/groovy/deploydb/DeployDBApp.groovy [hunk 2]
```diff
@@ -16,11 +16,12 @@ import org.joda.time.DateTimeZone
 import org.slf4j.Logger
 import org.slf4j.LoggerFactory
 
+
 class DeployDBApp extends Application<DeployDBConfiguration> {
     private final ImmutableList models = ImmutableList.of(
             models.Artifact, models.Deployment,
             models.PromotionResult, models.Flow)
-    private final Logger logger = LoggerFactory.getLogger(DeployDBApp.class)
+    private static final Logger logger = LoggerFactory.getLogger(DeployDBApp.class)
```
Comment (GiriDandu) [inline src/main/groovy/deploydb/DeployDBApp.groovy]: Why was this changed from final to static.  We have DeployDBApp available everywhere, why do you need to make this class scope?

File: src/main/groovy/deploydb/WorkFlow.groovy [hunk 3]
```diff
@@ -47,102 +55,295 @@ public class WorkFlow {
...
+        models.Webhook.Webhook tmpWebhook
+
+        /* Load promotions */
+        String promotionsDirName = baseConfigDirName + "/promotions"
+        File promotionsDirectory = new File(promotionsDirName);
+        if (promotionsDirectory.exists() && promotionsDirectory.isDirectory()) {
+            logger.info("Loading promotions from directory: ${promotionsDirectory.getCanonicalPath()}");
+
+            /* Skip everything but yaml file */
+            promotionsDirectory.eachFileMatch(FileType.FILES, ~/^.*?\.yml/) { File modelFile ->
```
Comment (GiriDandu) [inline src/main/groovy/deploydb/WorkFlow.groovy]: Can you not parameterize this functionality, this code appears few times below

File: src/main/groovy/deploydb/WorkFlow.groovy [hunk 4]
```diff
@@ -47,102 +55,295 @@ public class WorkFlow {
...
+                try {
+                    models.Promotion promotion = this.promotionLoader.load(modelFile)
+                    promotion.ident = this.promotionLoader.getIdent(modelFile.name)
+                    tmpPromotionRegistry.put(promotion.ident, promotion)
+                    logger.debug("Loaded promotions model: ${promotion.ident}")
+
+                } catch (all) {
+                    logger.info("Failed to load promotion model from ${modelFile.name}")
```
Comment (GiriDandu) [inline src/main/groovy/deploydb/WorkFlow.groovy]: In exception handling code, the model which are already configured will be initialized in half baked state. If this function is executed more than once, then there will be multiple entries and it is a problem. This is especially a big problem for webhooks, where we may send notifies more than once for a single configuration

File: src/test/groovy/deploydb/WorkFlowSpec.groovy [hunk 5]
```diff
@@ -13,3 +13,172 @@ class WorkFlowSpec extends Specification {
...
+        given:
+        createPromotionConfigFile()
+        createEnvironmentConfigFile()
+        createPipelineConfigFile()
+        createServiceConfigFile()
+        createWebhookConfigFile()
+        workFlow.loadConfigModels(baseCfgDirName)
+
+        expect:
+        workFlow.promotionRegistry.getAll().size() == 1
```
Comment (GiriDandu) [inline src/test/groovy/deploydb/WorkFlowSpec.groovy]: Why is only ident being checked here. Why not check the content of each registry.

File: deploydb.cucumber.yml (hunk 1 above)
Comment (maheshkelkar) [inline deploydb.cucumber.yml]: We dont need this config for cucumber test. But, I wondered if we need to leave this in there as an FYI

File: src/main/groovy/deploydb/DeployDBApp.groovy (hunk 2 above)
Comment (maheshkelkar) [inline src/main/groovy/deploydb/DeployDBApp.groovy]: I tried to make it consistent with usage in dropwizard app. Also, I dont think it is class scope, static will only guarantee only 1 instance for class

File: src/main/groovy/deploydb/WorkFlow.groovy (hunk 4 above)
Comment (maheshkelkar) [inline src/main/groovy/deploydb/WorkFlow.groovy]: We collect all of these models into temporary registry. Once we validate all the registries in the end and there are NO active flows, then we will overwrite the existing registries. So we are ensuring that there is no conflict

File: src/test/groovy/deploydb/WorkFlowSpec.groovy (hunk 5 above)
Comment (maheshkelkar) [inline src/test/groovy/deploydb/WorkFlowSpec.groovy]: We have test code that validates this in their respective spock tests. Here, we are only ensuring that expected number and ident model is populated in the application/workflow

File: src/main/groovy/deploydb/WorkFlow.groovy (hunk 4 above)
Comment (GiriDandu) [inline src/main/groovy/deploydb/WorkFlow.groovy]: Ah, I see in the code at the end of the function, you copy everything from temp registry to WorkFlow object, that should prevent duplicate entries. Although we have allocated objects in the process before exception handling, we expect the objects cleaned up by GC

File: src/test/groovy/deploydb/WorkFlowSpec.groovy (hunk 5 above)
Comment (GiriDandu) [inline src/test/groovy/deploydb/WorkFlowSpec.groovy]: Then I don't know why you need this test because we have the same identity tests in model spock test also. What's more this function is going to do?

**Comment by GiriDandu** [comment]
:+1:
